<a href="https://colab.research.google.com/github/KarolinaK-14/ML/blob/main/linear_regression_normal_equation.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

### scikit-learn
Library page: [https://scikit-learn.org](https://scikit-learn.org)

Documentation/User Guide: [https://scikit-learn.org/stable/user_guide.html](https://scikit-learn.org/stable/user_guide.html)

The core library for machine learning in Python.

To install the library, use the command below:
```
!pip install scikit-learn
```
To update the library to the latest version, use the command below:
```
!pip install --upgrade scikit-learn
```

### Table of contents:
1. [Importing libraries](#0)
2. [Preparing data](#1)
3. [Normal equation](#2)
4. [Linear regression using scikit-learn library](#3)




### <a name='0'></a> Importing libraries

In [1]:
# import numpy as np
import numpy as np

### <a name='1'></a> Preparing data

In [2]:
# X1 = np.array([1, 2, 3, 4, 5, 6])
# Y = np.array([3000, 3250, 3500, 3750, 4000, 4250])
# m = len(X1)

# print(f'Lata pracy: {X1}')
# print(f'Wynagrodzenie: {Y}')
# print(f'Liczba próbek: {m}')

X1 = np.array([1, 2, 3, 4, 5, 6])
Y = np.array([3000, 3250, 3500, 3750, 4000, 4250])
m = len(X1)

print(f'Years of experience: {X1}')
print(f'Salary: {Y}')
print(f'Number of samples: {m}')

Years of experience: [1 2 3 4 5 6]
Salary: [3000 3250 3500 3750 4000 4250]
Number of samples: 6


In [3]:
# X1 = X1.reshape(m, 1)
# print(X1)
# print(X1.shape)
print(X1.shape)
X1 = X1.reshape(m, 1)
print(X1.shape)
print(X1)

(6,)
(6, 1)
[[1]
 [2]
 [3]
 [4]
 [5]
 [6]]


In [4]:
# bias = np.ones((m, 1))
# print(bias)
# print(bias.shape)
bias = np.ones((m, 1))
print(bias.shape)
print(bias)

(6, 1)
[[1.]
 [1.]
 [1.]
 [1.]
 [1.]
 [1.]]


In [7]:
# X = np.append(bias, X1, axis=1)
# print(X)
# print(X.shape)
X = np.append(bias, X1, axis=1)
print(X.shape)
print(X)

(6, 2)
[[1. 1.]
 [1. 2.]
 [1. 3.]
 [1. 4.]
 [1. 5.]
 [1. 6.]]


### <a name='2'></a> Normal equation


Linear regression in $R^2$:  
$$Y = w_0 + w_1X_1$$
$$Y = W^TX$$
where: $$W = \left[\begin{matrix}w_0\\w_1\end{matrix}\right]$$ $$X = \left[\begin{matrix}1\\X_1\end{matrix}\right]$$ hence $$W^T = \left[\begin{matrix}w_0&w_1\end{matrix}\right]$$
$$Y = W^TX = \left[\begin{matrix}w_0&w_1\end{matrix}\right] \cdot \left[\begin{matrix}1\\X_1\end{matrix}\right] = w_0 + w_1X_1$$

Normal equation - the equation which allows to calculate the minimum of the loss function (if it exists)
$$W = (X^TX)^{-1}(X^TY)$$

In [ ]:
# np.dot(X.T, X)
np.dot(X.T, X)

array([[ 6., 21.],
       [21., 91.]])

In [8]:
L = np.linalg.inv(np.dot(X.T, X))

In [9]:
P = np.dot(X.T, Y)

In [10]:
np.dot(L, P)

array([2750.,  250.])

Final form of the model:$$Y = 2750 + 250X_1$$

### <a name='3'></a> Linear regression using sickit-learn library

In [13]:
# from sklearn.linear_model import LinearRegression

# regression = LinearRegression()
# regression.fit(X1, Y)

# print(regression.intercept_)
# print(regression.coef_[0])

from sklearn.linear_model import LinearRegression

regression = LinearRegression()
regression.fit(X1, Y)

print(regression.intercept_)
print(regression.coef_[0])

2750.0
250.0
